include/deal.II/differentiation/ad/ad_drivers.h

```diff
@@ -0,0 +1,1583 @@
+       *
+       * @return The scalar value of the function.
+       */
+      static ScalarType
+      value(const types::tape_index        active_tape_index,
+            const std::vector<ScalarType> &independent_variables);
+
+      /**
+       * Computes the gradient of the scalar field with respect to all
+       * independent variables.
+       *
+       * @param[in] active_tape_index The index of the tape on which the
+       *            dependent function is recorded.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] gradient The scalar values of the dependent function's
+       *             gradients. It is expected that this vector be of the
+       *             correct size (with length
+       *             <code>n_independent_variables</code>).
+       */
+      static void
+      gradient(const types::tape_index        active_tape_index,
+               const std::vector<ScalarType> &independent_variables,
+               Vector<ScalarType> &           gradient);
+
+      /**
+       * Computes the Hessian of the scalar field with respect to all
+       * independent variables.
+       *
+       * @param[in] active_tape_index The index of the tape on which the
+       *            dependent function is recorded.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] hessian The scalar values of the dependent function's
+       *             Hessian. It is expected that this matrix be of the correct
+       *             size (with dimensions
+       *             <code>n_independent_variables</code>$\times$<code>n_independent_variables</code>).
+       */
+      static void
+      hessian(const types::tape_index        active_tape_index,
+              const std::vector<ScalarType> &independent_variables,
+              FullMatrix<ScalarType> &       hessian);
+
+      //@}
+
+      /**
+       * @name Drivers for vector functions (multiple dependent variables)
+       */
+      //@{
+
+      /**
+       * Computes the values of the vector field.
+       *
+       * @param[in] active_tape_index The index of the tape on which the
+       *            dependent function is recorded.
+       * @param[in] n_dependent_variables The number of dependent variables.
+       * @param[in] independent_variables The scalar values of the independent
+       *            variables whose sensitivities were tracked.
+       * @param[out] values The scalar values of the dependent functions.
+       *             It is expected that this vector be of the correct size
+       *             (with length <code>n_dependent_variables</code>).
+       */
+      static void
+      values(const types::tape_index        active_tape_index,
+             const unsigned int             n_dependent_variables,
```

Comment:
I'm surprised you have to specify the number of dependent variables. Doesn't the function being taped know that? Or Maybe I just don't know (yet) how all of this is going to be used...

Reply:
> Doesn't the function being taped know that?

Somehow not... This is the downside to passing pointers rather than objects. I could infer this information from the size of the `values` vector, but for safety I check that its been initialised to the right size.